eventauth/eventcontent.go

```diff
@@ -0,0 +1,290 @@
+}
+
+// userIDAllowed checks whether the domain part of the user ID is allowed in
+// the room by the "m.federate" flag.
+func (c *createContent) userIDAllowed(id string) error {
+	domain, err := domainFromID(id)
+	if err != nil {
+		return err
+	}
+	return c.domainAllowed(domain)
+}
+
+// domainFromID returns everything after the first ":" character to extract
+// the domain part of a matrix ID.
+func domainFromID(id string) (string, error) {
+	// IDs have the format: SIGIL LOCALPART ":" DOMAIN
+	// Split on the first ":" character since the domain can contain ":"
+	// characters.
+	parts := strings.SplitN(id, ":", 2)
+	if len(parts) != 2 {
+		// The ID must have a ":" character.
+		return "", errorf("invalid ID: %q", id)
+	}
+	// Return everything after the first ":" character.
+	return parts[1], nil
+}
+
+// memberContent is the JSON content of a m.room.member event needed for
+// for auth checks.
+type memberContent struct {
+	Membership       string          `json:"membership"`
+	ThirdPartyInvite json.RawMessage `json:"third_party_invite"`
+}
+
+// Load the member content from the member event for the user ID.
+// Returns an error if there was an error loading the member event or
+// parsing the event content.
+func (c *memberContent) load(authEvents AuthEvents, userID string) error {
+	memberEvent, err := authEvents.Member(userID)
+	if err != nil {
+		return err
+	}
+	if memberEvent == nil {
+		// If there isn't a member event then the membership for the user
+		// defaults to leave.
+		c.Membership = "leave"
+		return nil
+	}
+	return c.parse(*memberEvent)
+}
+
+// Parse the member content of an event.
+// Returns an error if the content couldn't be parsed.
+func (c *memberContent) parse(event Event) error {
+	if err := json.Unmarshal(event.Content, c); err != nil {
+		return errorf("unparsable member event content: %s", err.Error())
+	}
+	return nil
+}
+
+// powerLevelContent is the JSON content of a m.room.power_levels event needed
+// for auth checks.
+// We can't unmarshal the content directly from JSON because we need to set
+// defaults and convert string values to int values.
+type powerLevelContent struct {
+	banLevel          int64
+	inviteLevel       int64
+	kickLevel         int64
+	redactLevel       int64
+	userLevels        map[string]int64
+	userDefaultLevel  int64
+	eventLevels       map[string]int64
+	eventDefaultLevel int64
+	stateDefaultLevel int64
+}
+
+// userLevel returns the power level a user has in the room.
+func (c *powerLevelContent) userLevel(userID string) int64 {
+	level, ok := c.userLevels[userID]
+	if ok {
+		return level
+	}
+	return c.userDefaultLevel
+}
+
+// eventLevel returns the power level needed to send an event in the room.
+func (c *powerLevelContent) eventLevel(eventType string, eventStateKey *string) int64 {
+	if eventType == "m.room.third_party_invite" {
+		// Special case third_party_invite events to have the same level as
+		// m.room.member invite events.
```

Comment:
As per [the spec](http://matrix.org/docs/spec/client_server/r0.2.0.html#m-room-third-party-invite) - this event is a state event but you do not assert this is the case.